hoodie-common/src/main/java/com/uber/hoodie/common/table/view/HoodieTableFileSystemView.java

```diff
@@ -186,47 +227,133 @@ private void writeObject(java.io.ObjectOutputStream out)
+        .filter(this::isDataFileNotDueToPendingCompaction)
         .map(Optional::get);
   }
 
   @Override
   public Stream<HoodieDataFile> getLatestDataFilesInRange(List<String> commitsToReturn) {
     return fileGroupMap.values().stream()
         .map(fileGroup -> fileGroup.getLatestDataFileInRange(commitsToReturn))
-        .filter(dataFileOpt -> dataFileOpt.isPresent())
+        .filter(this::isDataFileNotDueToPendingCompaction)
         .map(Optional::get);
   }
 
   @Override
   public Stream<HoodieDataFile> getAllDataFiles(String partitionPath) {
     return getAllFileGroups(partitionPath)
         .map(fileGroup -> fileGroup.getAllDataFiles())
-        .flatMap(dataFileList -> dataFileList);
+        .flatMap(dataFileList -> dataFileList)
+        .filter(this::isDataFileNotDueToPendingCompaction);
   }
 
   @Override
   public Stream<FileSlice> getLatestFileSlices(String partitionPath) {
     return getAllFileGroups(partitionPath)
         .map(fileGroup -> fileGroup.getLatestFileSlice())
-        .filter(dataFileOpt -> dataFileOpt.isPresent())
+        .filter(Optional::isPresent)
+        .map(Optional::get)
+        .map(this::unsetDataFileForFileSliceAfterPendingCompaction);
+  }
+
+  @Override
+  public Stream<FileSlice> getLatestUnCompactedFileSlices(String partitionPath) {
+    return getAllFileGroups(partitionPath)
+        .map(fileGroup -> {
+          FileSlice fileSlice = fileGroup.getLatestFileSlice().get();
+          // if the file-group is under compaction, pick the latest before compaction instant time.
+          if (isFileSliceAfterPendingCompaction(fileSlice)) {
+            String compactionInstantTime = fileIdToPendingCompactionInstantTime.get(fileSlice.getFileId());
+            return fileGroup.getLatestFileSliceBefore(compactionInstantTime);
+          }
+          return Optional.of(fileSlice);
+        })
         .map(Optional::get);
   }
 
+  /**
+   * Returns true if the file-group is under pending-compaction and the file-slice' baseInstant matches
+   * compaction Instant
+   * @param fileSlice File Slice
+   * @return
+   */
+  private boolean isFileSliceAfterPendingCompaction(FileSlice fileSlice) {
+    String compactionInstantTime = fileIdToPendingCompactionInstantTime.get(fileSlice.getFileId());
+    if ((null != compactionInstantTime) && fileSlice.getBaseInstantTime().equals(compactionInstantTime)) {
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * With concurrent compaction, it is possible to see partial/complete data-files due to inflight-compactions,
+   * Ignore those data-files
+   * @param fileSlice File Slice
+   * @return
+   */
+  private FileSlice unsetDataFileForFileSliceAfterPendingCompaction(FileSlice fileSlice) {
+    if (isFileSliceAfterPendingCompaction(fileSlice)) {
+      FileSlice transformed = new FileSlice(fileSlice.getBaseInstantTime(), fileSlice.getFileId());
+      fileSlice.getLogFiles().forEach(lf -> transformed.addLogFile(lf));
```

Comment:
add a comment explaining skipping of data file

Reply:
rename to something like "filterDataFileAfterPendingCompaction`

Reply:
Done